python_modules/dagster/dagster/_core/definitions/logical_version.py

```diff
@@ -0,0 +1,178 @@
+            ("input_logical_versions", Mapping["AssetKey", LogicalVersion]),
+        ],
+    )
+):
+    def __new__(
+        cls,
+        code_version: str,
+        input_logical_versions: Mapping["AssetKey", LogicalVersion],
+    ):
+        from dagster._core.definitions.events import AssetKey
+        return super(LogicalVersionProvenance, cls).__new__(
+            cls,
+            code_version=check.str_param(code_version, "code_version"),
+            input_logical_versions=check.mapping_param(
+                input_logical_versions,
+                "input_logical_versions",
+                key_type=AssetKey,
+                value_type=LogicalVersion,
+            ),
+        )
+
+
+# ########################
+# ##### TAG KEYS
+# ########################
+
+LOGICAL_VERSION_TAG_KEY: Final[str] = "dagster/logical_version"
+CODE_VERSION_TAG_KEY: Final[str] = "dagster/code_version"
+INPUT_LOGICAL_VERSION_TAG_KEY_PREFIX: Final[str] = "dagster/input_logical_version"
+INPUT_EVENT_POINTER_TAG_KEY_PREFIX: Final[str] = "dagster/input_event_pointer"
+
+
+def get_input_logical_version_tag_key(input_key: "AssetKey") -> str:
+    return f"f{INPUT_LOGICAL_VERSION_TAG_KEY_PREFIX}/{input_key.to_user_string()}"
+
+
+def get_input_event_pointer_tag_key(input_key: "AssetKey") -> str:
+    return f"{INPUT_EVENT_POINTER_TAG_KEY_PREFIX}/{input_key.to_user_string()}"
+
+
+# ########################
+# ##### COMPUTE / EXTRACT
+# ########################
+
+
+def compute_logical_version(
+    code_version: Union[str, UnknownValue], input_logical_versions: Sequence[LogicalVersion]
+) -> LogicalVersion:
+    """Compute a logical version from inputs.
+
+    Args:
+        code_version (LogicalVersion): The code version of the computation.
+        input_logical_versions (Sequence[LogicalVersion]): The logical versions of the inputs.
+
+    Returns:
+        LogicalVersion: The computed logical version.
+    """
+    check.inst_param(code_version, "code_version", (str, UnknownValue))
+    check.sequence_param(input_logical_versions, "input_versions", of_type=LogicalVersion)
+
+    if isinstance(code_version, UnknownValue) or UNKNOWN_LOGICAL_VERSION in input_logical_versions:
+        return UNKNOWN_LOGICAL_VERSION
+
+    all_inputs = [code_version, *(v.value for v in input_logical_versions)]
+
+    hash_sig = sha256()
+    hash_sig.update(bytearray("".join(all_inputs), "utf8"))
+    return LogicalVersion(hash_sig.hexdigest())
+
+
+@overload
+def extract_logical_version_from_event_log_entry(
+    event: EventLogEntry, *, include_provenance: Literal[False] = ...
+) -> Optional[LogicalVersion]:
+    ...
+
+
+@overload
+def extract_logical_version_from_event_log_entry(
+    event: EventLogEntry, *, include_provenance: Literal[True]
+) -> Tuple[Optional[LogicalVersion], Optional[LogicalVersionProvenance]]:
+    ...
+
+
+# Never set `include_provenance` to True when retrieving for a source asset.
+def extract_logical_version_from_event_log_entry(
```

Comment:
I am opposed to this because these properties only make sense for a small subset of `EventLogEntry` instances (it's a very general class).

Reply:
I've started adding properties like `asset_key` and `partition` to `EventLogRecord` - it just makes life way easier and doesn't really incur much cost.

That said, I don't feel super strongly about whether the methods should go on `EventLogRecord`.

I do think that the `include_provenance` arg is awkward and it would be clearer to separate it out into separate methods for extracting the version and provenance.